homeassistant/helpers/condition.py

```diff
@@ -56,6 +58,174 @@
+    """Append a TraceElement to trace[path]."""
+    trace = trace_var.get()
+    if trace is None:
+        trace_var.set({})
+        trace = trace_var.get()
+    trace.setdefault(path, [])
+    trace[path].append(trace_element)
+
+
+# Context variables for tracing
+# Config of condition being evaluated - TODO: Remove?
+condition_config = ContextVar("condition_config", default=None)
+# Trace of condition being evaluated
+condition_trace = ContextVar("condition_trace", default=None)
+# Stack of TraceElements
+condition_trace_stack = ContextVar("condition_trace_stack", default=None)
+# Current location in config tree
+condition_path_stack = ContextVar("condition_path_stack", default=None)
+
+
+def condition_trace_stack_push(node):
+    """Push a TraceElement to the top of the trace stack."""
+    trace_stack_push(condition_trace_stack, node)
+
+
+def condition_trace_stack_pop():
+    """Remove the top element from the trace stack."""
+    trace_stack_pop(condition_trace_stack)
+
+
+def condition_trace_stack_top():
+    """Return the element at the top of the trace stack."""
+    return trace_stack_top(condition_trace_stack)
+
+
+def condition_path_push(suffix):
+    """Go deeper in the config tree."""
+    if isinstance(suffix, str):
+        suffix = [suffix]
+    for node in suffix:
+        trace_stack_push(condition_path_stack, node)
+    return len(suffix)
+
+
+def condition_path_pop(n):
+    """Go n levels up in the config tree."""
+    for _ in range(n):
+        trace_stack_pop(condition_path_stack)
+
+
+def condition_path_get():
+    """Return a string representing the current location in the config tree."""
+    path = condition_path_stack.get()
+    if not path:
+        return ""
+    return "/".join(path)
+
+
+def condition_config_get():
+    """Return the config of the condition that was evaluated."""
+    return condition_config.get()
+
+
+def condition_trace_get():
+    """Return the trace of the condition that was evaluated."""
+    return condition_trace.get()
+
+
+def condition_trace_clear():
+    """Clear the condition trace."""
+    condition_config.set(None)
+    condition_trace.set(None)
+    condition_trace_stack.set(None)
+    condition_path_stack.set(None)
+
+
+def condition_trace_append(variables, path):
+    """Append a TraceElement to trace[path]."""
+    trace_element = TraceElement(variables)
+    trace_append_element(condition_trace, trace_element, path)
+    return trace_element
+
+
+def condition_trace_set_result(result, **kwargs):
+    """Set the result of TraceElement at the top of the stack."""
+    node = condition_trace_stack_top()
+
+    # The condition function may be called directly, in which case tracing
+    # is not setup
+    if not node:
+        return
+
+    node.set_result(result=result, **kwargs)
+
+
+@contextmanager
+def trace_condition(config, variables):
+    """Trace condition evaluation."""
+    if condition_config.get() is None:
+        condition_config.set(dict(config))
```

Comment:
Why make a copy? Configuration should never be mutated in place.

Reply:
Removed, there's no need to save the condition config, we only save full config of automation or script.